README.md
This is a vesy simple way to map your text data using [Altas from NOMIC](https://docs.nomic.ai/index.html) using the lib `click`.

You have to create an account to get API_KEY NOMIC. 

<< Atlas enables you to:
* Store, update and organize multi-million point datasets of unstructured text, images and embeddings.
* Visually interact with your datasets from a web browser.
* Run semantic search and vector operations over your datasets.
Use Atlas to:

    - Visualize, interact, collaborate and share large datasets of text and embeddings.
    - Collaboratively clean, tag and label your datasets
    - Build high-availability apps powered by semantic search
    - Understand and debug the latent space of your AI model trains  >>


# How to use
### Installation

To install the necessary dependencies, run the following command:

```bash
python -m venv mymapenv 
source mymapenv/bin/activate
pip install --upgrade pip 
pip install text2mapviewer
```

### Login NOMIC server

Login/create your Nomic account:

```bash
nomic login
```
If you have already your account : 

```bash
nomic login [YOUR_API_TOKEN_NOMIC_HERE] 
```
## Examples : 
#### from NOMIC and with lib text2mapviewer 
```python
from text2mapviewer.examples.map_embedding import project

# Use the projet from the lib text2mapviewer 
print(project) 
```

#### With the lib `click` after clone this ripo

```python
python scr/text2mapviewer/examples/map_embedding_click.py --num_embeddings 10000 --embedding_dim 256
```

#### [ The Animation Ouput](https://atlas.nomic.ai/map/0b4c0459-98f2-4aab-8d47-875765832049/54017477-907d-46e8-8d56-dddf7ab7fcfc)




## Supported Transformer Models from Hugging Face 

This project supports a variety of transformer models, including models from the Hugging Face Model Hub and sentence-transformers. Below are some examples:
    - Hugging Face Model: 'prajjwal1/bert-mini'
    - Hugging Face Model: 'Sahajtomar/french_semantic'  (french version for semantic search embedding) 
    - Sentence-Transformers Model: 'sentence-transformers/all-MiniLM-L6-v2' etc...

Please ensure that the model you choose is compatible with the project requirements and adjust the `--transformer_model_name` option accordingly.

## To map your text/csv  files

```bash
pip install -r requirements.txt
python main.py --transformer-model-name MODEL_NAME --cache_dir CACHE_DIR --batch-size BATCH_SIZE --file-path FILE_PATH
```
NOTE: for the CACHE_DIR : you can setup it like ==> 

```bash
export TRANSFORMERS_CACHE=/path_to_your/transformers_cache
```

Give a fidback. 


CHANGELOG.md
Change Log
==========

0.0.1 (14/04/2023)
-------------------
- First Release

fondamental/bases.py
# -- Papa Séga WADE 13/04/2023 --
from transformers import AutoTokenizer, AutoModel
import numpy as np
import torch
import csv
from typing import List, Dict,  Any, Optional, Tuple
from nomic import atlas

def load_documents(file_path: str) -> List[Dict[str, str]]:
    documents = []
    with open(file_path, 'r', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            documents.append({   
                """
                    Change with your own variables 
                """ 
                'filename': row['filename'],  
                'text': row['finalText'],
                'label': row['label'],
                'number_word': row['num_words']

            })
    return documents

def encode_documents(documents: List[Dict[str, str]], transformer_model_name: str, cache_dir: str, batch_size: int) -> np.ndarray:
    model = AutoModel.from_pretrained(transformer_model_name, cache_dir=cache_dir)
    tokenizer = AutoTokenizer.from_pretrained(transformer_model_name, cache_dir=cache_dir)
    embeddings = []

    with torch.no_grad():
        for i in range(0, len(documents), batch_size):
            batch = [document['text'] for document in documents[i:i+batch_size]]
            encoded_input = tokenizer(batch, return_tensors='pt', padding=True)
            cls_embeddings = model(**encoded_input)['last_hidden_state'][:, 0]
            embeddings.append(cls_embeddings)

    embeddings = torch.cat(embeddings).numpy()
    return embeddings

def map_embeddings_to_atlas(embeddings: np.ndarray, documents: List[Dict[str, str]]) -> str:
    response = atlas.map_embeddings(embeddings=embeddings, 
                                    data=documents,
                                    id_field='filename',
                                    colorable_fields=['label'],
                                    name="Name_Projet",
                                    description="Name_Project data map with Nomic Atlas.")
    return response


main.py
# -- Papa Séga WADE 13/04/2023 --
import click
import torch as th 
from typing import List, Dict
from fondamental.bases import * # or import --> load_documents, encode_documents, map_embeddings_to_atlas

@click.command()
@click.option('--transformer-model-name' , required=True, type=str, help='The name of the transformer model to load.')
@click.option('--cache_dir', envvar = 'TRANSFORMERS_CACHE', type=str, help='The directory to cache the transformer model in.')
@click.option('--batch-size', default=10, type=int, help='The batch size to use when encoding documents.')
@click.option('--file-path', default='dataset.csv', type=str, help='The path to the CSV file containing the documents.')
@click.pass_context
def main(ctx:click.core.Context, transformer_model_name: str, cache_dir: str, batch_size: int, file_path: str) -> None:
    device = th.device('cuda:0' if th.cuda.is_available() else 'cpu')
    ctx.obj = ctx.obj or {}
    ctx.obj['device'] = device
    documents = load_documents(file_path)
    ctx.obj['cache-dir'] = cache_dir
    embeddings = encode_documents(documents, transformer_model_name, cache_dir, batch_size)
    response = map_embeddings_to_atlas(embeddings, documents)
    print(response)

if __name__ == '__main__':
    main()

"""   To run the model :==> 
    python main.py --transformer-model-name MODEL_NAME --cache_dir CACHE_DIR --batch-size BATCH_SIZE --file-path FILE_PATH
    python main.py --transformer-model-name prajjwal1/bert-mini --cache_dir TRANSFORMERS_CACHE --batch-size 5 --file-path /path_to_file/dataset.csv

    models from hugginface
        -  model bert-mini : ==> prajjwal1/bert-mini
        -  model sentence_transformer ==> sentence-transformers/all-MiniLM-L6-v2
        -  model  BERT_Base  ==> distilbert-base-uncased
        - export TRANSFORMERS_CACHE=/home/pswia/miniconda3/envs/pswia/lib/python3.10/site-packages/transformers

"""
